asf_heat_pump_suitability/analysis/exploratory/create_full_dataset/assign_cluster_suitability_and_feasibility.py

```diff
@@ -0,0 +1,403 @@
+    df = df.with_columns(
+        (
+            (pl.col("cluster_size") - pl.col("cluster_size").min())
+            / (pl.col("cluster_size").max() - pl.col("cluster_size").min())
+            * 100
+        ).alias("cluster_size")
+    )
+
+    return df
+
+
+def calculate_feasibility_expression(tech_specific_weights: dict) -> pl.Expr:
+    """
+    Generates a Polars expression to calculate a weighted feasibility score.
+
+    Args:
+        weights_dict: A dictionary of the format {feature_name: weight}.
+
+    Returns:
+        pl.Expr: A Polars expression for the calculated score.
+    """
+    # Sum of all weights to normalize the score
+    total_weight = sum(tech_specific_weights.values())
+
+    # Compute weighted sum i.e. sum(feature_value * weight)
+    weighted_cols_sum = pl.sum_horizontal(
+        (
+            pl.col(f"perc_{feature}") * weight
+            if feature != "cluster_size"  # most features will have a "perc_" prefix
+            else pl.col("cluster_size")
+            * weight  # cluster_size does not have a "perc_" prefix
+        )
+        for feature, weight in tech_specific_weights.items()
+    )
+
+    return weighted_cols_sum / total_weight
+
+
+def create_df_feasibility_scoring(
+    df: pl.DataFrame,
+    features: list = config.features,
+    expected_tech_types: set = config.expected_tech_types,
+    weights: dict = config.weights,
+) -> pl.DataFrame:
+    """
+    Computes the feasibility score for each tech type and cluster in the dataset.
+
+    Tech types can include:
+    - Individual Air Source Heat Pump (ASHP)
+    - Collective purchasing of ASHP
+    - Shared Ground Loop (SGL)
+    - Heat Network (HN)
+
+    Args:
+        df (pl.DataFrame): DataFrame containing the dataset with feature values for each UPRN.
+        features (list, optional): List of features used in the feasibility scoring.
+            Defaults to a predefined list.
+        expected_tech_types (set, optional): Set of expected technology types for feasibility scoring.
+        weights (dict, optional): Dictionary containing the weights for each feature used to compute feasibility for specific tech types.
+            in the format {'tech_type_1': {'feature_1': weight_1, 'feature_2': weight_2, ...}, ..., 'tech_type_n': {...}}
+
+    Raises:
+        ValueError: if tech types or features provided in weights do not exist.
+
+    Returns:
+        pl.DataFrame: DataFrame with aggregated feasibility scores for each cluster.
+    """
+    if set(weights.keys()) != expected_tech_types:
+        incorrect_techs = [w for w in weights.keys() if w not in expected_tech_types]
+        missing_techs = [t for t in expected_tech_types if t not in weights.keys()]
+        raise ValueError(
+            f"There are incorrect or missing keys in the 'weights' dictionary.\n Incorrect techs: {incorrect_techs}\n Missing techs: {missing_techs}"
+        )
+
+    for tech in weights.keys():
+        tech_features = set(weights.get(tech).keys())
```

Comment:
```suggestion
        input_features = set(weights.get(tech).keys())
```
suggest renaming this variable so it's more readable :)